fsspec/caching.py

```diff
@@ -378,23 +417,27 @@ def _fetch(self, start, end):
             # First read, or extending both before and after
             self.cache = self.fetcher(start, bend)
             self.start = start
-        elif start < self.start:
-            if self.end - end > self.blocksize:
-                self.cache = self.fetcher(start, bend)
-                self.start = start
-            else:
-                new = self.fetcher(start, self.start)
-                self.start = start
-                self.cache = new + self.cache
-        elif bend > self.end:
-            if self.end > self.size:
-                pass
-            elif end - self.end > self.blocksize:
-                self.cache = self.fetcher(start, bend)
-                self.start = start
-            else:
-                new = self.fetcher(self.end, bend)
-                self.cache = self.cache + new
+        else:
+            assert self.start is not None
+            assert self.end is not None
```

Comment:
I needed to add this so that `mypy` wouldn't complain about `self.start` or `self.end` being None. Happy to restructure the code if you think there's a better way to do this.

Reply:
*THIS* annoys me about mypy. Asserts have no place outside of tests. The comparisons above would fail for the case of None.

Reply:
I would suggest `start = typing.cast(int, self.start)` and then using `start`.  Same for end.

This is also bunkum code which exists solely for the typechecker but at least the intention is a little clearer